### PR TITLE
[RFC] shell: add shell security module

### PR DESCRIPTION
**Pull request by vincent-d**
This PR adds a security layer for the shell.

It comes with a simple test app which shows the features.

The goal is to block the shell from unauthorized access. It is customizable and allow a multi level security access.

It can be improved and cleaned-up, but I'd like to have some feedback first.

### CODE REVIEW

File: tests/shell_sec/main.c [hunk 1]
```diff
@@ -0,0 +1,143 @@
...
+
+static char _user[10];
+
+static char user_list[NB_USER][USER_LENGTH] = {
+    "user",
+    "admin",
+};
+
+static char passwd_list[NB_USER][USER_LENGTH] = {
+    "passwd1",
```
Comment (aabadie) [inline tests/shell_sec/main.c]: what about using a list of encrypted passwords and compare the encrypted user entry with them ?

Comment (vincent-d) [inline tests/shell_sec/main.c]: Yes, this is just a proof of concept, it's not meant to be secured right now ;)

We can imagine a lot of different security and a basic one, user/password based, would use a hash or an encryption of the password I guess.

**Comment by miri64** [comment]
What is the use-case for this? Usually you wouldn't have the shell in a production environment, wouldn't you?

**Comment by bergzand** [comment]
This looks really nice!
So I might be going a bit too fast here, but this is an RFC so anyway. 
Would it not be better to have an interface to authentication modules? So that different authentication modules can provide different authentication methods which might be either required or optional. My idea borrows somewhat from the Unix PAM idea.

What I mean is that the login prompt could have an array of authentication modules. Each authentication module has a way to check the username/password and returns whether it is allowed. By marking these modules either as required or optional, some kind of logical function can be created.

I was personally thinking of looping the shell through (secured) CoAP to have a way to debug a node in a live environment and I think this would make a great addition to somewhat secure it. (I know that there are still many (security related) problems with this idea.)

**Comment by kYc0o** [comment]
> What is the use-case for this? Usually you wouldn't have the shell in a production environment, wouldn't you?

I think very cheap wall-powered border routers can be a good use case, maybe like the [Orion Router](http://zolertia.io/product/hardware/orion-router) (but far cheaper :P )

**Comment by vincent-d** [comment]
@miri64 In fact, I'd like to keep the shell in production. It could be handy at least in the first stages of our project. But then I need to limit the access.

@bergzand My idea was to make it possible for a given app to implement its own security. In our project we likely won't use a standard user/password authentication but a more advanced one. This proposal would make it possible to do so. But it could still be compatible with your idea I guess. A step by step approach sounds like a good idea to me ;)

Thanks for your comments!

**Comment by miri64** [comment]
@kYc0o wrote
> I think very cheap wall-powered border routers can be a good use case, maybe like the Orion Router (but far cheaper :P )

@vincent-d wrote
> @miri64 In fact, I'd like to keep the shell in production. It could be handy at least in the first stages of our project. But then I need to limit the access.

Intersting I always saw a RIOT-based border router more as a workaround, since anything more than a WRT-based solution seemed like a too big hassle for "mortal" users. But your use-cases sound legit :-)

**Comment by vincent-d** [comment]
I updated the proof of concept with a default security module which uses md5 hashes for passwords.

It now contains a list of user, associated to a level and a password (hashed).

**Comment by smlng** [comment]
as long is this tagged RFC, milestone seems premature.

**Comment by tcschmidt** [comment]
@aabadie doesn't this need another review round?

**Comment by smlng** [comment]
I would vote to close this, as is this is open for a while but didn't attract much attention. And also IMHO this is not the right way to go, but that's my personal opinion. Anyone else agreeing to close this?

**Comment by smlng** [comment]
I'm still suggesting to close this, see comments above.

Please add comments within the next days if you agree or object.

**Comment by smlng** [comment]
nothing new here, closing now

**Pull request closed without merging**